aptos-move/framework/aptos-framework/sources/jwks.move

```diff
@@ -0,0 +1,706 @@
+    public fun new_unsupported_jwk(id: vector<u8>, payload: vector<u8>): JWK {
+        JWK {
+            variant: copyable_any::pack(UnsupportedJWK { id, payload })
+        }
+    }
+
+    /// Initialize some JWK resources. Should only be invoked by genesis.
+    public(friend) fun initialize(account: &signer) {
+        system_addresses::assert_aptos_framework(account);
+        move_to(account, SupportedOIDCProviders { providers: vector[] });
+        move_to(account, ObservedJWKs { jwks: AllProvidersJWKs { entries: vector[] } });
+        move_to(account, Patches { patches: vector[] });
+        move_to(account, PatchedJWKs { jwks: AllProvidersJWKs { entries: vector [] } });
+    }
+
+    /// Only used by validators to publish their observed JWK update.
+    ///
+    /// NOTE: It is assumed verification has been done to ensure each update is quorum-certified,
+    /// and its `version` equals to the on-chain version + 1.
+    fun upsert_into_observed_jwks(aptos_framework: &signer, provider_jwks_vec: vector<ProviderJWKs>) acquires ObservedJWKs, PatchedJWKs, Patches {
+        system_addresses::assert_aptos_framework(aptos_framework);
+        let observed_jwks = borrow_global_mut<ObservedJWKs>(@aptos_framework);
+        vector::for_each(provider_jwks_vec, |obj| {
+            let provider_jwks: ProviderJWKs = obj;
+            upsert_provider_jwks(&mut observed_jwks.jwks, provider_jwks);
+        });
+
+        let epoch = reconfiguration::current_epoch();
+        emit(ObservedJWKsUpdated { epoch, jwks: observed_jwks.jwks });
+        regenerate_patched_jwks();
+    }
+
+    /// Regenerate `PatchedJWKs` from `ObservedJWKs` and `Patches` and save the result.
+    fun regenerate_patched_jwks() acquires PatchedJWKs, Patches, ObservedJWKs {
+        let jwks = borrow_global<ObservedJWKs>(@aptos_framework).jwks;
+        let patches = borrow_global<Patches>(@aptos_framework);
+        vector::for_each_ref(&patches.patches, |obj|{
+            let patch: &Patch = obj;
+            apply_patch(&mut jwks, *patch);
+        });
+        *borrow_global_mut<PatchedJWKs>(@aptos_framework) = PatchedJWKs { jwks };
+    }
+
+    /// Return whether a JWK can be found by issuer and key ID in a `AllProvidersJWKs`.
+    fun exists_in_all_providers_jwks(jwks: &AllProvidersJWKs, issuer: vector<u8>, jwk_id: vector<u8>): bool {
+        let (issuer_found, index) = vector::find(&jwks.entries, |obj| {
+            let provider_jwks: &ProviderJWKs = obj;
+            !is_greater_than(&compare_u8_vector(issuer, provider_jwks.issuer))
+        });
+
+        issuer_found && exists_in_provider_jwks(vector::borrow(&jwks.entries, index), jwk_id)
+    }
+
+    /// Return whether a JWK can be found by key ID in a `ProviderJWKs`.
+    fun exists_in_provider_jwks(provider_jwks: &ProviderJWKs, jwk_id: vector<u8>): bool {
+        vector::any(&provider_jwks.jwks, |obj| {
+            let jwk: &JWK = obj;
+            jwk_id == get_jwk_id(jwk)
```

Comment:
How come here you do not use `compare_u8_vector`?

Reply:
here the order doesn't matter.

Reply:
Wait, it's not about order... it's about correctly specifying an equality predicate. How come here you use == but in other places you [are supposed to] use `is_equal(compare_u8_vector())`?

Reply:
in the other places i'm supposed to exit the loop when i find the vector<u8> in the current position is larger than or equal to the target.

What i forgot to add is an additional
`issuer_found = (target == vector::borrow(vec, index))` after the loop.

Reply:
since the vector is ordered, i want to exit the loop early.

Reply:
I will try to fix everywhere to use `==` and try to understand the other places where you use `compare_u8_vector`.

Reply:
thanks! And for those `upsert_*` or `remove_*` functions where `is_greater_than` is used, they should be by design!